accepted/00000-ptf-repos-in-uyuni.md

```diff
@@ -0,0 +1,265 @@
+elements (incident 1234, first release):
+
+    Name: ptf-1234
+    Version: 1
+    Release: 1                 # always 1
+    Provides: ptf() = 1234-1
+    Requires: (pkg1 = pkg1EVR if pkg1)
+    Requires: (pkg2 = pkg1EVR if pkg2)
+    ...
+
+I.e. if pkg1 is installed it must be installed with version
+pkg1EVR. (We do not use Conflict: pkg1 != pkg1EVR because that
+would not work with multiversion packages like the kernel.)
+
+All packages providing ptf() get blacklisted by the solver,
+meaning they can only be installed by a specific solver job
+that addresses them. This means that selecting a specific
+ptf master package via yast or "zypper in" works, but they
+can not be pulled in via dependencies.
+
+Each individual rpm package must require the specific
+master ptf package:
+
+    Name: pkg1
+    Requires: ptf-1234 = 1-1
+
+This makes sure that the solver cannot pull in the rpm
+packages, as that would mean to also pull in the blacklisted
+master ptf package.
+
+
+### Updates
+
+Ptfs can be updated to a new release by calling 'zypper up'.
+This will update the master ptf package to some higher
+version (e.g. ptf-1234 = 2-1) and also pull in the corresponding
+rpm packages.
+
+
+### Uninstalling ptfs
+
+'zypper rm ptf-1234' will uninstall the ptf and revert back to
+non-ptf packages.
+
+
+### Making sure that no fixes are lost
+
+If all of the bugs fixed by a ptf are also fixed in maintenance
+updates, a new ptf release consisting only of the master ptf
+package is done:
+
+    Name: ptf-1234
+    Version: 3
+    Release: 1                 # always 1
+    Provides: self-destruct-pkg()
+    Requires: (pkg1 >= maintpkg1EVR if pkg1)
+    Requires: (pkg2 >= maintpkg2EVR if pkg2)
+
+Updating to this version will make sure that this system will
+contain only the fixed packages from the maintenance updates.
+The special "self-destruct-pkg()" provides will tell the
+solver that this will be a package erase instead of an
+install. This means that installing this package will actually
+erase the master ptf package.
+
+### Test Packages
+
+Before generating PTFs, SUSE will first create so called TEST packages.
+The customer need to test them and report positive result before SUSE make an official supported PTF out of it.
+
+TEST packages should only be installed on dedicated test systems.
+Therefor it is important that they get not installed on any system where just the version of the TEST package is higher than the installed one.
+
+
+## PTFs with Uyuni for SUSE and non SUSE OSes
+
+The general PTF design has 3 key points.
+
+1. A new ptf package to install the PTF.
+   
+   As this will be possible for SUSE OSes we need to check if such a package
+   can also be created for non SUSE OSes.
+   
+   - RHEL like OSes. As they also use rpm, it should be possible. We should check if RHEL 7 support the special kind of `Requires` dependencies and if yum and dnf behave like libzypp.
+   - Ubuntu and Debian. This needs some extra research if this is possible
```

Comment:
yes - but I have already an idea for a solution. I think in most cases we would talk about PTFs for salt and here we have a fix list of packages which must be installed. No need for checking "if" one of the subpackages is installed or not. We can just set hard requirements which works for every package manager.

But this is not really a question for this RFC as our maintenance team need to generate these packages.

We only care about the usability in Uyuni

Reply:
There will be also the Prometheus exporters, spacecmd and who knows what in the future. Maintaining a hardcoded list will be a burden. IMHO this can be better solved at the package level with `Dpkg::Pre-Invoke`. While the maintenance team will generate the packages, we should help them with non-SUSE OSs.

Reply:
I think there is a missunderstanding. The question is, how to do the following dependency in debian?

    Requires: (pkg1 = pkg1EVR if pkg1)

It means: if `pkg1` is installed, the ptf package has a depenency `Requires: pkg1 = pkg1EVR`. If pkg1 is not installed, this dependency does not exist. It is usefull when we create a PTF with multiple packages where some of them might not be installed (they are optional).
I think we do not have this situation with our limited number of packages. It could be sufficient to just specify the direct dependency without using this boolean dependency thingy.

Reply:
But that would be wrong. The goal is:

    Update all packages which are installed to a specific version,
    but ignore dependencies for packages which are not installed.

This is similar to how "patches" works.

Reply:
this topic was moved to a different rfc